File: code_samples/DataStructures/LinkedList/cs/SinglyLinkedList.cs
using System.Text;


// Обертка, выступающая в виде фиктивного узла
public class List<T> where T : IComparable<T>
{
    // Узел
    internal class Node 
    {
        public T? value; // Хранимое значение
        public Node? next; // Ссылка на следующий элемент цикла

        public Node(T? val)
        {
            value = val;
            next = null;
        }
    }
    internal Node? head; // указатель на первый элемент списка
    internal int length; // длина списка

    public int Length // свойство для доступа к значению длины
    {
        get => this.length;
    }

    public List()
    {
        this.head = null;
        this.length = 0;
    }

    // конструктор для составления списка из множества значений
    public List(params T?[] vals) : this()
    {
        if (vals is null || vals.Length == 0)
            return;

        this.head = new Node(vals[0]);
        Node? cur = this.head;

        for (var i = 1; i < vals.Length; i++) 
        {
            cur.next = new Node(vals[i]);
            cur = cur.next;
        }

        this.length = vals.Length;
    }

    // Добавление элемента в конец списка - медленно
    public List<T> Append(T? value)
    {
        this.length++;
        Node n = new(value);
        if (this.head is null)
        {
            this.head = n;
            return this;
        }

        Node current = this.head;
        while (current.next is not null)
            current = current.next;

        current.next = n;
        return this;
    }

    // Добавление элемента в начало списка - быстро
    public List<T> Prepend(T? value)
    {
        Node n = new(value);
        n.next = this.head;
        this.head = n;
        this.length++;

        return this;
    }

    // Добавление элемента в середину списка - зависит от позиции, но тоже медленно
    public List<T> AddAt(int idx, T? value)
    {
        if (idx < 0 || idx > this.Length)
                throw new System.IndexOutOfRangeException();

        if (idx == 0) return this.Prepend(value);

        if (idx == this.Length) return this.Append(value);

        Node cur = this.head;
        int j = 0;
        while (j < idx - 1)
        {
            cur = cur.next;
            j++;
        }

        Node nxt = new(value){next = cur.next};
        cur.next = nxt;
        this.length++;

        return this;
    }

    // Удаление элемента с начала списка
    public T? Shift()
    {
        if (this.head is null)
            return default(T);

        T result = this.head.value;
        this.head = this.head.next;
        this.length--;

        return result;
    }

    // Удаление элемента с конца списка
    public T? Pop()
    {
        if (this.head is null)
            return default(T);

        this.length--;
		
		if (this.head.next is null)
		{
			T result = this.head.value;
			this.head = null;
			return result;
		}

        Node prev = this.head;
		Node cur = prev.next;
        while (cur.next is not null) 
        {
            prev = cur;
            cur = cur.next;
        }

        prev.next = null;
        return cur.value;
    }

    public T? Delete(int idx)
    {
        if (idx < 0 || idx >= this.Length)
                throw new System.IndexOutOfRangeException();

        if (idx == 0) return this.Shift();

        Node prev = this.head;
		Node cur = prev.next;
        int j = 1;
        while (j < idx)
        {
            prev = cur;
            cur = cur.next;
            j++;
        }

        prev.next = cur.next;
        this.length--;

        return cur.value;
    }

    public T this[int idx]
    {
        get 
        {
            if (idx < 0 || idx >= this.length)
                throw new System.IndexOutOfRangeException();

            Node cur = this.head;
            int j = 0;
            while (j != idx)
            {
                cur = cur.next;
                j++;
            }

            return cur.value;
        }

        set
        {
            if (idx < 0 || idx >= this.length)
                throw new System.IndexOutOfRangeException();

            Node cur = this.head;
            int j = 0;
            while (j != idx)
            {
                cur = cur.next;
                j++;
            }

            cur.value = value;
        }
    }

    public T? Head() => this.head.value;

    public List Tail() => new List(){
        length = this.Length - 1,
        head = this.head.next
    };

    public int Find(T val)
    {
        Node cur = this.head;
        for (int i = 0; i < this.Length; i++)
        {
            if (cur.value.CompareTo(val) == 0)
                return i;

            cur = cur.next;
        }

        return -1;
    }

    public override string ToString()
    {
        if (this.head is null) return "[]";

        StringBuilder sb = new("[");

        Node current = this.head;

        while (current.next is not null)
        {
            sb.AppendFormat("{0}, ", current.value);
            current = current.next;
        }

        return sb.AppendFormat("{0}]", current.value).ToString();
    }
}

File: code_samples/DataStructures/LinkedList/go/singly_linked_list.go
package list

import (
	"fmt"
	"strings"
)

// node - узел связанного списка
type node[T comparable] struct {
	value T
	next  *node[T] // указатель на следующий элемент списка
}

// List - обертка/фиктивный узел списка
type List[T comparable] struct {
	length int
	head   *node[T] // указатель на голову - первый элемент - списка
}

// New - конструктор списка
func New[T comparable](params ...T) *List[T] {
	list := &List[T]{length: len(params)}

	if list.length == 0 {
		return list
	}

	list.head = &node[T]{value: params[0]}
	cur := list.head
	for i := 1; i < len(params); i++ {
		cur.next = &node[T]{value: params[i]}
		cur = cur.next
	}

	return list
}

// Prepend - вставка элемента в начало списка - долго
func (l *List[T]) Prepend(value T) *List[T] {
	n := &node[T]{
		value: value,
		next:  l.head,
	}
	l.head = n
	l.length++

	return l
}

// Append - вставка элемента в конец списка - медленно
func (l *List[T]) Append(value T) *List[T] {
	n := &node[T]{value: value}
	l.length++

	if l.head == nil {
		l.head = n
		return l
	}

	var cur *node[T]
	for cur = l.head; cur.next != nil; cur = cur.next {
	}

	cur.next = n
	return l
}

// AddAt - вставка элемента в середину списка - тоже медленно
func (l *List[T]) AddAt(idx int, value T) (*List[T], error) {
	if idx < 0 && idx > l.length {
		return nil, fmt.Errorf("index %d is out of range: list length is %d", idx, l.length)
	}

	if idx == 0 {
		return l.Prepend(value), nil
	}

	if idx == l.length {
		return l.Append(value), nil
	}

	cur := l.head
	for j := 0; j < idx-1; j++ {
		cur = cur.next
	}

	n := &node[T]{
		value: value,
		next:  cur.next,
	}

	cur.next = n
	l.length++
	return l, nil
}

// Shift - удаление и возвращение элемента из начала списка
func (l *List[T]) Shift() T {
	result := l.head.value
	l.head = l.head.next
	l.length--

	return result
}

// Pop - удаление и возвращение элемента из конца списка
func (l *List[T]) Pop() T {
	if l.head == nil {
		var val T
		return val
	}

	if l.head.next == nil {
		val := l.head.value
		l.head = nil
		return val
	}

	prev := l.head
	cur := prev.next
	for cur.next != nil {
		prev = cur
		cur = cur.next
	}

	prev.next = nil
	l.length--
	return cur.value
}

// Delete - удаление и возвращение элемента по индексу
func (l *List[T]) Delete(idx int) (T, error) {
	if idx < 0 && idx >= l.length {
		var val T
		return val, fmt.Errorf("index %d is out of range: list length is %d", idx, l.length)
	}

	if idx == 0 {
		return l.Shift(), nil
	}

	prev := l.head
	cur := prev.next
	for j := 1; j < idx; j++ {
		prev = cur
		cur = cur.next
	}

	prev.next = cur.next
	l.length--
	return cur.value, nil
}

// Get - получение элемента по индексу
func (l List[T]) Get(idx int) (T, error) {
	if idx < 0 && idx >= l.length {
		var val T
		return val, fmt.Errorf("index %d is out of range: list length is %d", idx, l.length)
	}

	cur := l.head
	for cur.next != nil {
		cur = cur.next
	}

	return cur.value, nil
}

// Set - установка элемента по индексу
func (l List[T]) Set(idx int, value T) error {
	if idx < 0 && idx >= l.length {
		return fmt.Errorf("index %d is out of range: list length is %d", idx, l.length)
	}

	cur := l.head
	for cur.next != nil {
		cur = cur.next
	}

	cur.value = value
	return nil
}

// Head - получение значения первого элемента списка
func (l *List[T]) Head() T {
	return l.head.value
}

// Tail - получение хвоста списка
func (l *List[T]) Tail() *List[T] {
	return &List[T]{
		length: l.length - 1,
		head:   l.head.next,
	}
}

// Find - поиск позиции первого совпадающего элемента
func (l *List[T]) Find(value T) int {
	cur := l.head
	for j := 0; j < l.length; j++ {
		if cur.value == value {
			return j
		}

		cur = cur.next
	}

	return -1
}

func (l List[T]) String() string {
	sb := new(strings.Builder)
	sb.WriteRune('[')

	for cur := l.head; cur != nil; cur = cur.next {
		sb.WriteString(fmt.Sprintf("%v,", cur.value))
	}

	return strings.TrimRight(sb.String(), ",") + "]"
}
